lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -220,112 +229,177 @@ define((require, exports, module) ->
-
-      curSequenceIndex = EquationRangeHandler.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
-      curEquationStart = null
-      while curSequenceIndex >= 0
-        curToken = tokenIterator.stepBackward()
-        if not curToken
-          return null
-        if EquationRangeHandler.equalTokens(
-            EquationRangeHandler.BEGIN_EQUATION_TOKEN_SEQUENCE[curSequenceIndex],
-            curToken)
-          if curSequenceIndex == EquationRangeHandler.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
-            curTokenPosition = tokenIterator.getCurrentTokenPosition()
-            curEquationStart = {
-              row: curTokenPosition.row
-              column: curTokenPosition.column + curToken.value.length
-            }
-          curSequenceIndex -= 1
-        else
-          curSequenceIndex = EquationRangeHandler.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
-          curEquationStart = null
-      return curEquationStart
-
-    getEquationEnd: (tokenIterator) ->
-      # if tokenIterator is initially on the empty line, its current token is null
-      if not tokenIterator.getCurrentToken()?
-        tokenIterator.stepBackward()
-        # if tokenIterator is still null, then we're at the end of a file
-        if not tokenIterator.getCurrentToken()?
-          return null
-      else
-        # following cycle pushes tokenIterator to the start of
-        # ending sequence, if it is inside one
-        # The loop isn't executed, if current token is null, because:
-        #   a) we don't need to -- if `tokenIterator` is initially on
-        #      the empty line, then it is guaranteed not to be inside start sequence
-        #   b) it can cause bugs -- if `tokenIterator` is initially after the end of a document,
-        #      current token is null, and after stepping backward `tokenIterator` is on the
-        #      end of a document. If end sequence happens to be there, then equation end is
-        #      then found without any problem, whereas in this case null should be returned
-        for token in EquationRangeHandler.END_EQUATION_TOKEN_SEQUENCE.slice(0).reverse()
-          if EquationRangeHandler.equalTokens(token, tokenIterator.getCurrentToken())
-            tokenIterator.stepBackward()
-
-      curSequenceIndex = 0
-      curEquationStart = null
-      while curSequenceIndex < EquationRangeHandler.END_EQUATION_TOKEN_SEQUENCE.length
-        curToken = tokenIterator.stepForward()
-        if not curToken
+        for boundarySequence in boundarySequences
+          for token in boundarySequence.slice(0).reverse()
+            curToken = tokenIterator.getCurrentToken()
+            if curToken? and EquationRangeHandler.equalTokens(token, curToken)
+              moveFromBoundary()
+
+      allowedSequences = Array(boundarySequences.length)
+      allowedSequences.fill(true)
+      curSequence = null
+      curSequenceId = null
+      curSequenceIndex = null
+      curEquationOuterBoundary = null
```

Comment:
pls add a comment explaining the difference between inner and outer boundaries (when they can be different?)